store/sqldb/migrate.go

```diff
@@ -418,3 +463,278 @@ func (s *SpDBImpl) DeleteMigrateGVGUnitsByBucketID(bucketID uint64) error {
 	}
 	return nil
 }
+
+func (s *SpDBImpl) UpdateBucketMigrationProgress(bucketID uint64, migrateState int) error {
+	var (
+		result              *gorm.DB
+		insertMigrateBucket *MigrateBucketProgressTable
+		queryReturn         *MigrateBucketProgressTable
+		needInsert          bool
+	)
+	queryReturn = &MigrateBucketProgressTable{}
+	insertMigrateBucket = &MigrateBucketProgressTable{
+		BucketID:       bucketID,
+		MigrationState: migrateState,
+	}
+	result = s.db.First(queryReturn, "bucket_id = ?", bucketID)
+	if result.Error != nil && !errors.Is(result.Error, gorm.ErrRecordNotFound) {
+		return result.Error
+	}
+	if result.Error != nil {
+		needInsert = errors.Is(result.Error, gorm.ErrRecordNotFound)
+	}
+
+	if needInsert {
+		result = s.db.Create(insertMigrateBucket)
+		if result.Error != nil || result.RowsAffected != 1 {
+			return fmt.Errorf("failed to insert record in migrate bucket table: %s", result.Error)
+		}
+	} else { // update
+		result = s.db.Model(&MigrateBucketProgressTable{}).
+			Where("bucket_id = ?", bucketID).Updates(insertMigrateBucket)
+		if result.Error != nil {
+			return fmt.Errorf("failed to update record in migrate bucket table: %s", result.Error)
+		}
+	}
+
+	return nil
+}
+
+func (s *SpDBImpl) QueryMigrateBucketState(bucketID uint64) (int, error) {
+	var (
+		result              *gorm.DB
+		insertMigrateBucket *MigrateBucketProgressTable
+		queryReturn         *MigrateBucketProgressTable
+		needInsert          bool
+		state               = 0
+	)
+	queryReturn = &MigrateBucketProgressTable{}
+	insertMigrateBucket = &MigrateBucketProgressTable{
+		BucketID:       bucketID,
+		MigrationState: state,
+	}
+	result = s.db.First(queryReturn, "bucket_id = ?", bucketID)
+	if result.Error != nil && !errors.Is(result.Error, gorm.ErrRecordNotFound) {
+		return state, result.Error
+	}
+	if result.Error != nil {
+		needInsert = errors.Is(result.Error, gorm.ErrRecordNotFound)
+	}
+
+	if needInsert {
+		result = s.db.Create(insertMigrateBucket)
+		if result.Error != nil || result.RowsAffected != 1 {
+			return 0, fmt.Errorf("failed to insert record in migrate bucket table: %s", result.Error)
+		}
+	} else {
+		state = queryReturn.MigrationState
+	}
+
+	return state, nil
+}
+
+func (s *SpDBImpl) QueryMigrateBucketProgress(bucketID uint64) (*spdb.MigrateBucketProgressMeta, error) {
+	var (
+		result      *gorm.DB
+		queryReturn *MigrateBucketProgressTable
+	)
+
+	progress := &spdb.MigrateBucketProgressMeta{}
+
+	queryReturn = &MigrateBucketProgressTable{}
+	result = s.db.First(queryReturn, "bucket_id = ?", bucketID)
+	if result.Error != nil && !errors.Is(result.Error, gorm.ErrRecordNotFound) {
+		return &spdb.MigrateBucketProgressMeta{}, result.Error
+	}
+	if errors.Is(result.Error, gorm.ErrRecordNotFound) {
```

Comment:
What is the special meaning of this if? Which has been included in the revious if condition